### PR TITLE
XRay comparison v2

### PR DESCRIPTION
**Pull request by kdoh**
Includes the ability to compare segments to other segments and growth rate information for time series cards.

### CODE REVIEW

**Comment by attekei** [comment]
I couldn't get past the loading spinner in any of xrays pages.

## Frontend errors

If I try to enter `/xray/table/1/approximate` I get an unhelpful React error:
```
ReactReconciler.js?abcd:67 Uncaught (in promise) TypeError: Cannot read property 'getHostNode' of null
    at Object.getHostNode (ReactReconciler.js?abcd:67)
```

If I create any question with `Count of rows` aggregation and date breakout using a table from Sample Dataset or Pies table I get a remapping-related frontend error:
```
Uncaught (in promise) TypeError: Cannot read property 'map' of undefined
    at extractRemappedColumns (index.js?f04e:88)
    at index.js?f04e:82
    at Array.map (<anonymous>)
    at extractRemappings (index.js?f04e:77)
    at Visualization.transform (Visualization.jsx?7bd2:197)
    at Visualization.componentWillMount (Visualization.jsx?7bd2:127)
```

## Backend errors

Question with `Sad Toucan Incidents table / Count of rows / Timestamp: Week` causes a server error when entering XRays:

```
{:message "java.lang.String cannot be cast to java.lang.Number",
 :stacktrace
 ["feature_extraction.histogram$histogram.invokeStatic(histogram.clj:14)"
  "feature_extraction.histogram$histogram.invoke(histogram.clj:10)"
  "feature_extraction.core$dataset__GT_features.invokeStatic(core.clj:29)"
  "feature_extraction.core$dataset__GT_features.invoke(core.clj:26)"
  "feature_extraction.core$eval50289$fn__50290.invoke(core.clj:83)"
  "api.x_ray$fn__50357$fn__50358.invoke(x_ray.clj:76)"
  "api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:230)"
  "api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:225)"
  "api.x_ray$fn__50357.invokeStatic(x_ray.clj:69)"
  "api.x_ray$fn__50357.invoke(x_ray.clj:69)"
  "middleware$enforce_authentication$fn__41250.invoke(middleware.clj:122)"
  "api.routes$fn__50433.invokeStatic(routes.clj:65)"
  "api.routes$fn__50433.invoke(routes.clj:65)"
  "routes$fn__50505$fn__50506.doInvoke(routes.clj:75)"
  "routes$fn__50505.invokeStatic(routes.clj:71)"
  "routes$fn__50505.invoke(routes.clj:71)"
  "middleware$log_api_call$fn__41365$fn__41367.invoke(middleware.clj:331)"
  "middleware$log_api_call$fn__41365.invoke(middleware.clj:330)"
  "middleware$add_security_headers$fn__41299.invoke(middleware.clj:245)"
  "middleware$bind_current_user$fn__41254.invoke(middleware.clj:142)"
  "middleware$maybe_set_site_url$fn__41303.invoke(middleware.clj:268)"]
```

When trying to show a time series question in xrays (don't remember the exact query atm – I think I was using Orders table in Pies and a month breakout):
```
08-24 11:25:43 ERROR metabase.middleware :: GET /api/x-ray/card/95 500 (716 ms) (101 DB calls)
{:message "Value 13 for monthOfYear must be in the range [1,12]",
 :stacktrace
 ["feature_extraction.feature_extractors$round_to_month.invokeStatic(feature_extractors.clj:459)"
  "feature_extraction.feature_extractors$round_to_month.invoke(feature_extractors.clj:455)"
  "feature_extraction.feature_extractors$month_frequencies.invokeStatic(feature_extractors.clj:463)"
  "feature_extraction.feature_extractors$month_frequencies.invoke(feature_extractors.clj:461)"
  "feature_extraction.feature_extractors$eval50112$fn__50114$fn__50116.invoke(feature_extractors.clj:515)"
  "feature_extraction.feature_extractors$eval50112$fn__50114.invoke(feature_extractors.clj:506)"
  "feature_extraction.core$x_ray$fn__50303.invoke(core.clj:119)"
  "util$update_when.invokeStatic(util.clj:866)"
  "util$update_when.doInvoke(util.clj:862)"
  "feature_extraction.core$x_ray.invokeStatic(core.clj:116)"
  "feature_extraction.core$x_ray.invoke(core.clj:110)"
  "api.x_ray$fn__50357$fn__50358.invoke(x_ray.clj:74)"
  "api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:230)"
  "api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:225)"
  "api.x_ray$fn__50357.invokeStatic(x_ray.clj:69)"
  "api.x_ray$fn__50357.invoke(x_ray.clj:69)"
  "middleware$enforce_authentication$fn__41250.invoke(middleware.clj:122)"
  "api.routes$fn__50433.invokeStatic(routes.clj:65)"
  "api.routes$fn__50433.invoke(routes.clj:65)"
  "routes$fn__50505$fn__50506.doInvoke(routes.clj:75)"
  "routes$fn__50505.invokeStatic(routes.clj:71)"
  "routes$fn__50505.invoke(routes.clj:71)"
  "middleware$log_api_call$fn__41365$fn__41367.invoke(middleware.clj:331)"
  "middleware$log_api_call$fn__41365.invoke(middleware.clj:330)"
  "middleware$add_security_headers$fn__41299.invoke(middleware.clj:245)"
  "middleware$bind_current_user$fn__41254.invoke(middleware.clj:142)"
  "middleware$maybe_set_site_url$fn__41303.invoke(middleware.clj:268)"]}
```

This I got for many simple queries (count of rows, date breakout) in Sample Dataset and Pies:
```
{:message nil,
 :stacktrace
 ["feature_extraction.feature_extractors$weigh_periodicity$fn__50108.invoke(feature_extractors.clj:485)"
  "feature_extraction.core$trim_decimals.invokeStatic(core.clj:101)"
  "feature_extraction.core$trim_decimals.invoke(core.clj:99)"
  "feature_extraction.core$x_ray$fn__50303.invoke(core.clj:119)"
  "util$update_when.invokeStatic(util.clj:866)"
  "util$update_when.doInvoke(util.clj:862)"
  "feature_extraction.core$x_ray.invokeStatic(core.clj:116)"
  "feature_extraction.core$x_ray.invoke(core.clj:110)"
  "api.x_ray$fn__50357$fn__50358.invoke(x_ray.clj:74)"
  "api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:230)"
  "api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:225)"
  "api.x_ray$fn__50357.invokeStatic(x_ray.clj:69)"
  "api.x_ray$fn__50357.invoke(x_ray.clj:69)"
  "middleware$enforce_authentication$fn__41250.invoke(middleware.clj:122)"
  "api.routes$fn__50433.invokeStatic(routes.clj:65)"
  "api.routes$fn__50433.invoke(routes.clj:65)"
  "routes$fn__50505$fn__50506.doInvoke(routes.clj:75)"
  "routes$fn__50505.invokeStatic(routes.clj:71)"
  "routes$fn__50505.invoke(routes.clj:71)"
  "middleware$log_api_call$fn__41365$fn__41367.invoke(middleware.clj:331)"
  "middleware$log_api_call$fn__41365.invoke(middleware.clj:330)"
  "middleware$add_security_headers$fn__41299.invoke(middleware.clj:245)"
  "middleware$bind_current_user$fn__41254.invoke(middleware.clj:142)"
  "middleware$maybe_set_site_url$fn__41303.invoke(middleware.clj:268)"]}
```

**Comment by attekei** [comment]
If you use a segment in your question, you see the xrays option even if the query doesn't have a time breakout:

screen shot 2017-08-24 at 13 18 04

**Comment by kdoh** [comment]
@attekei That's expected. If you're using segments in your question we want people to be able to xray those segments.

**Comment by attekei** [comment]
@kdoh Okay, I think I also encountered some odd errors with that one which I'll try to reproduce again after the problems with `extractRemappedColumns` have been resolved (now it seems to shadow all other errors).

File: frontend/src/metabase/routes.jsx [hunk 1]
```diff
@@ -254,17 +255,27 @@ export const getRoutes = (store) =>
...
+                        <Route path="segments/:segmentId1/:segmentId2">
+                            <Route path=":cost" component={SegmentComparison} />
+                            <Route path="field/:fieldName/:cost" component={SegmentFieldComparison} />
+                        </Route>
+                        <Route path="segment/:segmentId/table/:tableId">
+                            <Route path=":cost" component={SegmentTableComparison} />
+                            <Route path="field/:fieldName/:cost" component={SegmentFieldComparison} />
+                        </Route>
+                        { /* NYI */ }
+                        <Route paht="fields/:fieldId1/:fieldId2" component={FieldComparison} />
```
Comment (attekei) [inline frontend/src/metabase/routes.jsx]: `paht`=>`path`

File: frontend/src/metabase/xray/components/Periodicity.jsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import Histogram from 'metabase/xray/Histogram'
+
+const Periodicity = ({ xray }) =>
+    <div>
+        <Heading heading="Time breakdown" />,
+        <div className="bg-white bordered rounded shadowed">
+            <div className="Grid Grid--gutters Grid--1of4">
+                { PERIODICITY.map(period =>
+                    xray[`histogram-${period}`] && (
+                        <div className="Grid-cell">
```
Comment (attekei) [inline frontend/src/metabase/xray/components/Periodicity.jsx]: there could be a Period component for rendering each period

File: frontend/src/metabase/xray/selectors.js [hunk 3]
```diff
@@ -21,14 +24,92 @@ export const getSegmentConstituents = (state) =>
...
 
-export const getComparison = (state) => state.xray.comparison
+export const getComparison = (state) => state.xray.comparison && state.xray.comparison
+
+export const getComparisonFields = createSelector(
+    [getComparison],
+    (comparison) => {
+        if(comparison) {
+            return Object.keys(comparison.constituents[0].constituents)
+                             .map(key => {
```
Comment (attekei) [inline frontend/src/metabase/xray/selectors.js]: pretty weird indentation here

File: frontend/src/metabase/xray/selectors.js [hunk 4]
```diff
@@ -21,14 +24,92 @@ export const getSegmentConstituents = (state) =>
...
-                     })
+export const getComparisonContributors = createSelector(
+    [getComparison],
+    (comparison) => {
+        if(comparison) {
+            const top = comparison['top-contributors']
+            return top && top.map(contributor => {
+                return Object.keys(comparison.constituents[0].constituents)
+                      .map(key => {
+                          if(key === contributor.field.toUpperCase()) {
```
Comment (attekei) [inline frontend/src/metabase/xray/selectors.js]: I don't completely understand this – this seems to return `undefined` every time the current key doesn't match the contributor field, but those `undefined`s are never filtered out at least in this selector

**Review by attekei — COMMENTED** [review]
Left some small notes, generally looks good!

**Comment by mazameli** [comment]
A bunch of tiny notes, most of which shouldn't hold up merging. I was viewing this on my laptop, so a lot of the visual comments I think are about the way things look when the viewport is around 1200px wide.

- I can get segment x-rays to load, but not time series if they use a Count of Rows (from the action button on a saved time series question). As Atte mentioned, the loader just keeps spinning.
- The teal and purple on the comparison pages looks nice, but do we use the teal anywhere else? Kind of looks like a color palette from a different product.
- Do we want x-ray to be the first option?

screen shot 2017-08-24 at 3 22 02 pm

- Is it expected that I have to save a question that uses a segment in order to x-ray it?
- Probably a problem for another day, but one funny UX thing is that, when I'm looking at an x-ray of a segment (like "Large Orders"), and then click on a field in the segment, the field x-ray I'm taken to is the unsegmented data for that field (i.e., it's no longer "filtered" by that segment), so the values change.
- When I click on a field in a table or segment in an x-ray, my scroll position isn't reset to the top of the screen.
- When navigating through several different x-rays, I find myself getting a little lost as to what kind of thing I'm looking at: table/segment vs. field. Don't know if we want to add a label somewhere near the title that says what kind of entity this is.
- The `fidelity` options don't look to me like they're vertically aligned with any other elements, and they seem slightly big to me:

screen shot 2017-08-24 at 3 56 27 pm

- Small pluralization error:
screen shot 2017-08-24 at 3 56 36 pm

- `Maximal value` could use more left margin so that the words don't run together:
screen shot 2017-08-24 at 3 56 54 pm

- Not sure I'm digging the arrangement of these elements. It feels like a breadcrumb, but it isn't really one. Maybe it would be simpler for the title to just say, "X-ray of [entity]". (It should be `X-ray` rather than `XRay`, too.)
screen shot 2017-08-24 at 4 01 09 pm

- The spacing on this grid is really tight at the 1200px-ish breakpoint:
screen shot 2017-08-24 at 4 01 24 pm

- Not sure what the class is that's setting the letter-spacing here, but seems a bit tight, and is missing a comma:
screen shot 2017-08-24 at 4 02 14 pm

- Could use some padding at the bottom of this table:
screen shot 2017-08-24 at 4 08 43 pm

- Not totally sure what's going on with this table, but I assume it's because the range is long and the values aren't very tall. What are the circles for? (This was comparing a segment called "Large Orders" to all of Orders.)

- This column feels hard to scan. I wonder if these various "ratings" could be represented as a scale, like each rating could be represented as a row of 10 skinny rectangles that are filled in depending on the rating on a scale of 1–10.

screen shot 2017-08-24 at 4 09 33 pm

- Not a huge deal, but I didn't realize these were clickable for a while:
screen shot 2017-08-24 at 4 18 48 pm

- It would be cool if the day-of-week value could be the name of the day instead of the number:
screen shot 2017-08-24 at 4 21 46 pm

- Same comment as above, but for months:
screen shot 2017-08-24 at 4 21 58 pm

- Seems overly bold/contrasty:
screen shot 2017-08-24 at 4 22 34 pm

**Comment by sbelak** [comment]
I think I've fixed all 3 BE issues now.

**Pull request merged**